torch/distributed/optim/utils.py

```diff
@@ -0,0 +1,41 @@
+from typing import Type
+from torch import optim
+from .functional_adagrad import _FunctionalAdagrad
+from .functional_adam import _FunctionalAdam
+from .functional_adamw import _FunctionalAdamW
+from .functional_sgd import _FunctionalSGD
+from .functional_adadelta import _FunctionalAdadelta
+from .functional_rmsprop import _FunctionalRMSprop
+from .functional_rprop import _FunctionalRprop
+from .functional_adamax import _FunctionalAdamax
+
+# dict to map a user passed in optimizer_class to a functional
+# optimizer class if we have already defined inside the
+# distributed.optim package, this is so that we hide the
+# functional optimizer to user and still provide the same API.
+functional_optim_map = {
+    optim.Adagrad: _FunctionalAdagrad,
+    optim.Adam: _FunctionalAdam,
+    optim.AdamW: _FunctionalAdamW,
+    optim.SGD: _FunctionalSGD,
+    optim.Adadelta: _FunctionalAdadelta,
+    optim.RMSprop: _FunctionalRMSprop,
+    optim.Rprop: _FunctionalRprop,
+    optim.Adamax: _FunctionalAdamax,
+}
+
+def as_functional_optim(optim_cls: Type, *args, **kwargs):
+    try:
+        functional_cls = functional_optim_map[optim_cls]
+    except KeyError:
+        raise ValueError(f"Optimizer {optim_cls} does not have a functional counterpart!")
+
+    return create_functional_optim(functional_cls, *args, **kwargs)
+
+def create_functional_optim(functional_optim_cls: Type, *args, **kwargs):
+    return functional_optim_cls(
+        [],
+        *args,
+        **kwargs,
+        _allow_empty_param_list=True,
```

Comment:
Same comment as in #71602. I think `_allow_empty_param_list` condition should be handled inside `_OptimizerHookState`.

Reply:
The main issue is that `_OptimizerHookState` will expect a fully constructed functional optimizer, so the caller of that needs to have constructed a functional optimizer which requires passing in _allow_empty_param_list with the rest of the arguments. `create_functional_optim` / `as_functional_optim` eventually becomes quite hidden as we only call it in `_OverlappedOptimizer` which is further called from DDP. 

If we want to specify allow_empty_param_list inside ``_OptimizerHookState`, there are 2 options I can think of:

1. Construct the entire functional optimizer inside `_OptimizerHookState` 
2. Just omit `_allow_empty_param_list` arg here, and then manually set it in `_OptimizerHookState`. This hides the arg a bit better but seems hackier as a developer will have to look in 2 places to see how exactly the functional optimizer is being constructed.